File: multi_language_code_generator/implementations/javascript/log_analyzer.js
class LogAnalyzer {
    parseLine(line) {
        // Splits by arbitrary whitespace
        const parts = line.trim().split(/\s+/);
        
        if (parts.length === 5 && parts[1] === '-') {
            const statusCode = parseInt(parts[4], 10);
            if (!isNaN(statusCode)) {
                return {
                    ip_address: parts[0],
                    method: parts[2],
                    endpoint: parts[3],
                    status_code: statusCode
                };
            }
        }
        return null;
    }

    analyze(lines) {
        let total_requests = 0;
        const unique_visitors_set = new Set();
        let errors = 0;

        for (const line of lines) {
            const parsed = this.parseLine(line);
            if (parsed !== null) {
                total_requests += 1;
                unique_visitors_set.add(parsed.ip_address);
                if (parsed.status_code >= 400) {
                    errors += 1;
                }
            }
        }

        let error_rate = 0.0;
        if (total_requests > 0) {
            error_rate = Math.round((errors / total_requests) * 100) / 100;
        }

        return {
            total_requests: total_requests,
            unique_visitors: unique_visitors_set.size,
            error_rate: error_rate
        };
    }
}

module.exports = LogAnalyzer;

File: multi_language_code_generator/implementations/go/log_analyzer.go
package log_analyzer

import (
	"math"
	"strconv"
	"strings"
)

type LogEntry struct {
	IPAddress  string
	Method     string
	Endpoint   string
	StatusCode int
}

type AnalysisResult struct {
	TotalRequests  int     `json:"total_requests"`
	UniqueVisitors int     `json:"unique_visitors"`
	ErrorRate      float64 `json:"error_rate"`
}

type LogAnalyzer struct{}

func NewLogAnalyzer() *LogAnalyzer {
	return &LogAnalyzer{}
}

func (a *LogAnalyzer) ParseLine(line string) *LogEntry {
	parts := strings.Fields(line)
	if len(parts) == 5 && parts[1] == "-" {
		statusCode, err := strconv.Atoi(parts[4])
		if err == nil {
			return &LogEntry{
				IPAddress:  parts[0],
				Method:     parts[2],
				Endpoint:   parts[3],
				StatusCode: statusCode,
			}
		}
	}
	return nil
}

func (a *LogAnalyzer) Analyze(lines []string) AnalysisResult {
	totalRequests := 0
	uniqueVisitorsSet := make(map[string]struct{})
	errors := 0

	for _, line := range lines {
		parsed := a.ParseLine(line)
		if parsed != nil {
			totalRequests++
			uniqueVisitorsSet[parsed.IPAddress] = struct{}{}
			if parsed.StatusCode >= 400 {
				errors++
			}
		}
	}

	errorRate := 0.0
	if totalRequests > 0 {
		errorRate = float64(errors) / float64(totalRequests)
		errorRate = math.Round(errorRate*100) / 100
	}

	return AnalysisResult{
		TotalRequests:  totalRequests,
		UniqueVisitors: len(uniqueVisitorsSet),
		ErrorRate:      errorRate,
	}
}
